net: ocpp: add mixed array support for ocpp1.6 JSON messages

SanjayyyV opened the pull request:
Replaced the manual string parsing to parse mixed array JSON messages with the native mixed array JSON support in JSON library

Signed-off-by: Sanjay Vallimanalan <sanjay@linumiz.com>

Fixes: #112424

SanjayyyV committed:
net: ocpp: add mixed array support for ocpp1.6 JSON messages

Replaced the manual string parsing to parse mixed array JSON messages
with the native mixed array JSON support in JSON library

Signed-off-by: Sanjay Vallimanalan <sanjay@linumiz.com>

ssekar15 commented on subsys/net/lib/ocpp/ocpp_j.c:
```diff
@@ -473,57 +435,73 @@ static int frame_status_resp_msg(char *buf, int len, char *res, char *uid)
+		strncpy(uid, call_res_msg.uid, uidlen - 1);
+		uid[uidlen - 1] = '\0';
+
+		p_len = strlen(call_res_msg.payload);
```
ditto

SanjayyyV commented on subsys/net/lib/ocpp/ocpp_j.h:
```diff
@@ -25,6 +25,21 @@
 #define SAMPLED_VALUE_MIN_FIELDS 2
 #define SAMPLED_VALUE_MAX_FIELDS 3
 
+struct json_ocpp_call_req_msg {
+	uint8_t call_req;
+	char *uid;
+	const char *action;
+	char *payload;
+	uint8_t count;
+};
+
+struct json_ocpp_call_res_msg {
+	uint8_t call_res;
+	char *uid;
+	char *payload;
+	uint8_t count;
+};
```
@jukkar This comment has been addressed. Kindly review at your earliest convenience.

jukkar replied:
I do not see my comment addressed

SanjayyyV replied:
@jukkar Addressed. The original 3-byte saving assumed count was uint8_t; since it had to become size_t for the encode-side count_offset contract, that packing no longer applies. Reshuffled to group the word-sized members together with the lone uint8_t last:
```
struct json_ocpp_call_req_msg {
	char *uid;
	const char *action;
	char *payload;
	size_t count;
	uint8_t call_req;
};

struct json_ocpp_call_res_msg {
	char *uid;
	char *payload;
	size_t count;
	uint8_t call_res;
};
```

jukkar commented:
This PR went under my review radar, sorry about that.

SanjayyyV commented:
@jukkar Comments are addressed. Kind ping for review.

jukkar commented on subsys/net/lib/ocpp/ocpp_j.c:
```diff
@@ -473,57 +435,70 @@ static int frame_status_resp_msg(char *buf, int len, char *res, char *uid)
 	if (msg == NULL || uid == NULL || pdu == NULL || is_rsp == NULL) {
 		return -EINVAL;
 	}
 
```
This looks like a proper check suggestion

SanjayyyV replied:
@jukkar Addressed by adding this check to the existing set of checks.

jukkar commented on subsys/net/lib/ocpp/ocpp_j.c:
```diff
@@ -473,57 +435,70 @@ static int frame_status_resp_msg(char *buf, int len, char *res, char *uid)
-	}
-
+	rpc_id = msg[1] - '0';
 	switch (rpc_id + '0') {
```
@SanjayyyV can you check the copilot comments, if they are false positives, please comment that, otherwise please fix the issues.

SanjayyyV replied:
@jukkar Addressed, removed the redundant `rpc_id`.

SanjayyyV committed:
tests: net: ocpp: add unit test for parse_rpc_msg

The OCPP RPC parser was switched to json_mixed_arr_parse(), but the
only existing test exercises it indirectly through a live central
system and is build-only, so the parse path had no automated coverage.

Add a ztest that feeds representative OCPP 1.6 frames directly into
parse_rpc_msg() and checks the parsed uid, payload and response flag
for both Call and CallResult. Truncated frames are also fed in to
verify they are rejected rather than parsed with unset fields.

Signed-off-by: Sanjay Vallimanalan <sanjay@linumiz.com>

SanjayyyV commented on subsys/net/lib/ocpp/ocpp_j.c:
```diff
@@ -473,57 +435,70 @@ static int frame_status_resp_msg(char *buf, int len, char *res, char *uid)
+
+		strncpy(msg, call_req_msg.payload, msglen - 1);
+		msg[msglen - 1] = '\0';
+		break;
```
Addressed by scoping each case statement with proper set of braces.

SanjayyyV commented on subsys/net/lib/ocpp/ocpp_j.c:
```diff
@@ -473,65 +435,82 @@ static int frame_status_resp_msg(char *buf, int len, char *res, char *uid)
-	rpc_id = *token - '0';
+		ret = json_mixed_arr_parse(msg, msglen, call_req_descr,
+					   ARRAY_SIZE(call_req_descr), &call_req_msg);
+		if (ret < 0) {
```
Addressed by adding a ZTEST.

SanjayyyV commented on subsys/net/lib/ocpp/ocpp_j.c:
```diff
@@ -473,65 +435,82 @@ static int frame_status_resp_msg(char *buf, int len, char *res, char *uid)
-	if (ret < 0) {
-		return ret;
+		strncpy(msg, call_req_msg.payload, msglen - 1);
+		msg[msglen - 1] = '\0';
```
Fixed. Since payload aliases into msg, switched the in-place copy to memmove sized by strlen(payload) + 1 (payload + NUL), which is overlap-safe. The following memset clears the leftover tail so no stale bytes remain past the terminator.
